### PR TITLE
Issue 3: User Model & Authentication

### LINKED ISSUE #3: User Model & Authentication
### Goal

Implement **user registration and authentication** using JWT. This is a core part of the MVP — users must be able to register, log in, and receive an authentication token to access protected endpoints.

---

### Description

1. **Generate User model**

   * Fields: `name`, `email`, `password_digest`, `role`.
   * Add model validations:

     * `name`: presence.
     * `email`: presence, uniqueness, format validation.
     * `password_digest`: required.
   * Use `has_secure_password` (requires `bcrypt` gem).

2. **JWT Setup**

   * Add `jwt` gem (if not already included).
   * Create a service class `JwtService` to:

     * Encode payloads with a secret.
     * Decode tokens and handle expiration.
   * Store secret key in `Rails.application.credentials` or `.env`.

3. **Auth Controller**

   * `POST /auth/register` → create new user, return JWT.
   * `POST /auth/login` → authenticate user (email + password), return JWT.
   * Return JSON response: `{ token: <jwt>, user: {...} }`.

4. **Middleware / Concern for Authentication**

   * Create `ApplicationController#authorize_request` method.
   * Decode JWT from `Authorization: Bearer <token>` header.
   * Set `@current_user` for protected controllers.

5. **Testing**

   * Unit test: `User` model validations.
   * Request tests for `/auth/register` and `/auth/login`.
   * Edge cases:

     * Invalid email/password.
     * Duplicate email on registration.

6. **Documentation**

   * Update Swagger/OpenAPI spec:

     * `POST /auth/register` (request body + response).
     * `POST /auth/login`.

---

### Deliverables

* `User` model with validations and password hashing.
* `AuthController` with `register` and `login` endpoints.
* JWT service implemented and integrated.
* Authentication middleware/concern for future use.
* Passing RSpec tests for model + endpoints.
* API docs updated.

---

### Dependencies

* Requires **Task 2 (Database Schema Draft)** to define `users` table.
* Can be developed in parallel with **Task 4 (Event model & API)** since they don’t depend directly on each other.

### LINKED ISSUE #32: Roles & UserRoles Models
* **Goal:** Implement a proper role system matching the new ER diagram.
* **Pre‑requisites:** The existing `User` model and authentication setup should already be working.
* **Steps:**

  1. Generate `Role` model/table with fields: `id`, `name` (varchar(64), unique, not null).
  2. Generate `UserRole` join model/table (`user_roles`) with fields: `user_id`, `role_id` and a composite primary key on both columns.
  3. Remove the simple `role` string from `User` (if present) and add associations: `User has_many :roles, through: :user_roles`.
  4. Seed default roles (`user`, `admin`).
  5. On user registration, create a `user_role` record assigning the default role (`user`).
  6. Update JWT payload to include an array of role names.
  7. Write RSpec tests for model associations and validations.

### PR DESCRIPTION
**Pull request by PavloKilko**

### CODE REVIEW

**Comment by PavloKilko** [comment]
I suggest adding Swagger API documentation for the user auth after merging the branch with the MkDocs setup into the develop branch.

**Comment by SomeTyraredGuy** [comment]
Controllers should be under api/v1/ to implement API versioning

```
api/					# Rails API root
├───app/				# Application code
|	├───controllers/    # REST API controllers, organized by resource
|	|	└───api/        # API namespace for versioning
|	|		└───vN/     # Version N of the API (e.g., v1, v2)
```

File: api/app/controllers/auth_controller.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        refresh_token: tokens[:refresh_token],
+        user: {
+          id: user.id,
+          name: user.name,
+          email: user.email,
+          roles: user.roles.select(:id, :name)
+        }
+      }, status: :created
+    else
+      render json: { errors: user.errors.full_messages }, status: :unprocessable_entity
```
[the remark this answers is not in the document]
Comment (PavloKilko) [reply api/app/controllers/auth_controller.rb]: fixed.

File: api/app/models/user_role.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class UserRole < ApplicationRecord
+  self.primary_key = :user_id, :role_id
```
[the remark this answers is not in the document]
Comment (PavloKilko) [reply api/app/models/user_role.rb]: fixed.

File: api/db/migrate/20250923122136_create_users.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[8.0]
+  def change
+    create_table :users do |t|
+      t.string :name, null: false, limit: 64
+      t.string :email, null: false, limit: 128
+      t.string :password_digest, null: false, limit: 64
```
[the remark this answers is not in the document]
Comment (PavloKilko) [reply api/db/migrate/20250923122136_create_users.rb:6]: The bcrypt password hashing algorithm produces a hash with a fixed length, regardless of the input password's length. So I suggest keeping 64 character length limit.

File: api/app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class UsersController < ApplicationController
```
[the remark this answers is not in the document]
Comment (PavloKilko) [reply api/app/controllers/users_controller.rb]: No, before_action callbacks run parent class first, then child

File: api/app/controllers/concerns/authenticatable.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    token = header.start_with?("Bearer") ? header.split(" ", 2).last : nil
+    raise JwtService::InvalidToken, "Missing bearer token" if token.blank?
+
+    payload = JwtService.decode(token)
+    @current_user = User.find(payload[:sub])
+  end
+
+  def require_role!(*role_names)
+    unless current_user&.roles&.exists?(name: role_names)
+      render json: { error: "Forbidden" }, status: :forbidden
```
[the remark this answers is not in the document]
Comment (PavloKilko) [reply api/app/controllers/concerns/authenticatable.rb]: return is redundant

File: api/spec/requests/auth_spec.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      expect(json["user"]["email"]).to eq("jane@example.com")
+      expect(json["user"]["roles"]).to be_an(Array)
+      expect(json["user"]["roles"].map { |r| r["name"] }).to include("USER")
+    end
+
+    it "fails on duplicate email" do
+      User.create!(name: "Old", email: "dup@example.com", password: "pass123")
+      post "/auth/register", params: { name: "New", email: "dup@example.com", password: "pass123" }, as: :json
+
+      expect(response).to have_http_status(:unprocessable_content)
```
[the remark this answers is not in the document]
Comment (PavloKilko) [reply api/spec/requests/auth_spec.rb]: fixed

**Comment by PavloKilko** [comment]
> Controllers should be under api/v1/ to implement API versioning
> 
> ```
> api/					# Rails API root
> ├───app/				# Application code
> |	├───controllers/    # REST API controllers, organized by resource
> |	|	└───api/        # API namespace for versioning
> |	|		└───vN/     # Version N of the API (e.g., v1, v2)
> ```

fixed

File: api/app/controllers/concerns/authenticatable.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    end
+  end
+
+  def current_user
+    @current_user
+  end
+
+  def authorize_request
+    header = request.headers["Authorization"].to_s
+    token = header.start_with?("Bearer") ? header.split(" ", 2).last : nil
```
Comment (mehalyna) [inline api/app/controllers/concerns/authenticatable.rb]: The Authorization header parsing does not handle "bearer" in lowercase, which is valid per RFC 6750. Update the header parsing to be case-insensitive.
```
header = request.headers["Authorization"].to_s
token = header[/\ABearer\s+/i] ? header.split(" ", 2).last : nil
```

File: api/app/controllers/concerns/authenticatable.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    raise JwtService::InvalidToken, "Missing bearer token" if token.blank?
+
+    payload = JwtService.decode(token)
+    @current_user = User.find(payload[:sub])
+  end
+
+  def require_role!(*role_names)
+    unless current_user&.roles&.exists?(name: role_names)
+      render json: { error: "Forbidden" }, status: :forbidden
+    end
```
Comment (mehalyna) [inline api/app/controllers/concerns/authenticatable.rb:29]: The require_role! method does not halt execution after rendering a forbidden error, which may cause double render errors. Add a return statement after rendering the forbidden error.
```
render json: { error: "Forbidden" }, status: :forbidden
return
```

File: api/app/services/jwt_service.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    JWT.encode(body, secret_key, ALGORITHM)
+  end
+
+  def self.decode(token)
+    decoded, = JWT.decode(token, secret_key, true, { algorithm: ALGORITHM })
+    decoded.with_indifferent_access
+  rescue JWT::ExpiredSignature
+    raise ExpiredToken, "Token has expired"
+  rescue JWT::DecodeError
+    raise InvalidToken, "Token is invalid"
```
Comment (mehalyna) [inline api/app/services/jwt_service.rb]: The decode method does not handle all potential JWT errors, such as JWT::VerificationError. Rescue additional JWT errors to avoid unhandled exceptions.

```
rescue JWT::ExpiredSignature
  raise ExpiredToken, "Token has expired"
rescue JWT::DecodeError, JWT::VerificationError
  raise InvalidToken, "Token is invalid"
```

File: api/app/services/jwt_service.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  class ExpiredToken < StandardError; end
+  class InvalidToken < StandardError; end
+
+  def self.secret_key
+    ENV.fetch("JWT_SECRET_KEY")
+  end
+
+  def self.encode(payload, exp:)
+    body = payload.dup
+    body[:exp] = exp.to_i
```
Comment (mehalyna) [inline api/app/services/jwt_service.rb]: The method does not ensure the payload is a Hash with symbol keys, which may cause inconsistencies. Ensure payload keys are symbols before encoding.
```
body = payload.dup.symbolize_keys
body[:exp] = exp.to_i
```

File: api/app/models/user_role.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+class UserRole < ApplicationRecord
+  self.primary_key = [ :user_id, :role_id ]
```
Comment (mehalyna) [inline api/app/models/user_role.rb]: As far as I remember, this setting is not necessary for Rails join tables and may cause issues.

File: api/db/migrate/20250923122523_create_user_roles.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+class CreateUserRoles < ActiveRecord::Migration[8.0]
+  def change
+    create_table :user_roles, id: false, primary_key: %i[user_id role_id] do |t|
+      t.references :user, null: false, foreign_key: true, index: false
+      t.references :role, null: false, foreign_key: true, index: false
+      t.timestamps
+    end
+    add_index :user_roles, %i[user_id role_id], unique: true
```
Comment (mehalyna) [inline api/db/migrate/20250923122523_create_user_roles.rb]: Creating the user_roles table with id: false and primary_key: is non-standard and may cause migration issues. You may omit the primary_key option and use a composite unique index instead.
```
create_table :user_roles, id: false do |t|
  t.references :user, null: false, foreign_key: true, index: false
  t.references :role, null: false, foreign_key: true, index: false
  t.timestamps
end
add_index :user_roles, [:user_id, :role_id], unique: true
```

File: api/app/controllers/api/v1/users_controller.rb [hunk 13]
```diff
@@ -0,0 +1,11 @@
+module Api
+  module V1
+    class UsersController < BaseController
+      before_action -> { require_role!("USER") }, only: :me
+
+      def me
+        render json: { id: current_user.id, name: current_user.name, email: current_user.email }
```
Comment (mehalyna) [inline api/app/controllers/api/v1/users_controller.rb]: The me action does not return user roles, which may be needed for frontend authorization. Include roles in the response for consistency.
```
render json: { id: current_user.id, name: current_user.name, email: current_user.email, roles: current_user.roles.select(:id, :name) }
```

File: api/spec/requests/api/v1/auth_spec.rb [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    end
+  end
+
+  describe "POST /api/v1/auth/login" do
+    let!(:user) { User.create!(name: "Joe", email: "joe@example.com", password: "password123") }
+
+    it "returns tokens for valid credentials" do
+      post "/api/v1/auth/login", params: { email: "joe@example.com", password: "password123" }, as: :json
+
+      expect(response).to have_http_status(:ok)
```
Comment (mehalyna) [inline api/spec/requests/api/v1/auth_spec.rb]: The test does not check for case-insensitive emails, which the login code supports. Add a test for login with mixed-case email.
```
post "/api/v1/auth/login", params: { email: "JOE@EXAMPLE.COM", password: "password123" }, as: :json
expect(response).to have_http_status(:ok)
```

File: api/app/services/jwt_service.rb [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  rescue JWT::ExpiredSignature
+    raise ExpiredToken, "Token has expired"
+  rescue JWT::DecodeError
+    raise InvalidToken, "Token is invalid"
+  end
+
+  def self.issue_tokens_for(user)
+    now = Time.current
+    access = encode({ sub: user.id, typ: "access" }, exp: now + ACCESS_TTL)
+    refresh = encode({ sub: user.id, typ: "refresh" }, exp: now + REFRESH_TTL)
```
Comment (mehalyna) [inline api/app/services/jwt_service.rb]: The issue_tokens_for method does not include jti (JWT ID), which can help with token revocation. Add a unique jti claim to the access and refresh tokens.
```
jti = SecureRandom.uuid
access = encode({ sub: user.id, typ: "access", jti: jti }, exp: now + ACCESS_TTL)
refresh = encode({ sub: user.id, typ: "refresh", jti: jti }, exp: now + REFRESH_TTL)
```

File: api/.env.example [hunk 16]
```diff
@@ -1 +1,6 @@
+# This is example of the environment configuration parameters
+
+# DO NOT USE EXACT VALUES IN PRODUCTION
+# DO NOT MODIFY WITHOUT PERMISSION
 DATABASE_URL=postgres://postgres:postgres@localhost:6000/app_development
+JWT_SECRET_KEY=jwtsecretkey
```
Comment (le-murier) [inline api/.env.example:6]: Non-blocking: could use rails credentials, but it depends on how team want to share secrets/deploy/store them
https://thoughtbot.com/blog/switching-from-env-files-to-rails-credentials

File: api/app/controllers/concerns/authenticatable.rb (hunk 8 above)
Comment (PavloKilko) [reply api/app/controllers/concerns/authenticatable.rb:29]: I don’t think that non-use of return statements are a big issue here, since all `render json: {...}` calls are already inside `if/else/unless` blocks. Also, our customized RuboCop setup will flag this as a style violation `[Style/RedundantError]`.

File: api/app/controllers/concerns/authenticatable.rb (hunk 7 above)
Comment (PavloKilko) [reply api/app/controllers/concerns/authenticatable.rb]: fixed

File: api/app/services/jwt_service.rb (hunk 9 above)
Comment (PavloKilko) [reply api/app/services/jwt_service.rb]: fixed

File: api/app/services/jwt_service.rb (hunk 10 above)
Comment (PavloKilko) [reply api/app/services/jwt_service.rb]: fixed

File: api/app/models/user_role.rb (hunk 11 above)
Comment (PavloKilko) [reply api/app/models/user_role.rb]: fixed

File: api/db/migrate/20250923122523_create_user_roles.rb (hunk 12 above)
Comment (PavloKilko) [reply api/db/migrate/20250923122523_create_user_roles.rb]: fixed

File: api/app/controllers/api/v1/users_controller.rb (hunk 13 above)
Comment (PavloKilko) [reply api/app/controllers/api/v1/users_controller.rb]: Added, but I suggest treating this endpoint as experimental for now, and in the future moving it into a dedicated controller with a different API scope (e.g. `/api/v1/account*` with endpoints as `/me`, `/change-password`, `/recovery-password`, etc.).

File: api/app/services/jwt_service.rb (hunk 15 above)
Comment (PavloKilko) [reply api/app/services/jwt_service.rb]: fixed

File: api/.env.example (hunk 16 above)
Comment (PavloKilko) [reply api/.env.example:6]: Duly noted, I'll discuss it with the team.

File: api/spec/requests/api/v1/auth_spec.rb (hunk 14 above)
Comment (PavloKilko) [reply api/spec/requests/api/v1/auth_spec.rb]: Added

File: api/app/controllers/api/v1/auth_controller.rb [hunk 17]
```diff
@@ -0,0 +1,57 @@
+module Api
+  module V1
+    class AuthController < BaseController
+      skip_before_action :authorize_request, only: %i[register login]
+
+      def register
+        user = User.new(register_params)
+
+        if user.save
+          user.roles << Role.find_by!(name: "USER")
```
Comment (le-murier) [inline api/app/controllers/api/v1/auth_controller.rb]: in rails we generally extract things into service objects or serializers, so controller do less, not more
https://medium.com/@thilorusche/service-objects-for-rails-9c5973dc8bc2
https://medium.com/@stvik2/serializing-data-in-rails-a46cb16e0c2d

Comment (le-murier) [inline api/app/controllers/api/v1/auth_controller.rb]: I assume this is temporary: `user.roles << Role.find_by!(name: "USER")`
Just don't forget to update it later, as role can be renamed and etc., maybe drop a comment in code about it

File: api/app/controllers/api/v1/auth_controller.rb [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+              roles: user.roles.select(:id, :name)
+            }
+          }, status: :created
+        else
+          render json: { errors: user.errors.full_messages }, status: :unprocessable_entity
+        end
+      end
+
+      def login
+        user = User.find_by("LOWER(email) = ?", login_params[:email].to_s.downcase)
```
Comment (le-murier) [inline api/app/controllers/api/v1/auth_controller.rb]: could use ILIKE, but main issue that it is performed in controller, if app needs case-insensitive logins, then why are we not enforcing this at the DB level?

Comment (le-murier) [reply api/app/controllers/api/v1/auth_controller.rb]: we can normalize emails to lowercase on save, or use a case-insensitive index

File: api/app/controllers/api/v1/auth_controller.rb (hunk 17 above)
Comment (le-murier) [reply api/app/controllers/api/v1/auth_controller.rb]: you can:

- extract user creation + role assignment into a UserRegistrationService 
- extract token generation into a TokenResponseBuilder or include it in the JwtService
- use ActiveModel::Serializer or just a manual serializer for the user response.

File: api/app/controllers/api/v1/auth_controller.rb [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+          }, status: :ok
+        else
+          render json: { error: "Invalid email or password" }, status: :unauthorized
+        end
+      end
+
+      private
+
+      def register_params
+        params.require(:name)
```
Comment (le-murier) [inline api/app/controllers/api/v1/auth_controller.rb:25]: https://api.rubyonrails.org/v7.0.7/classes/ActionController/Parameters.html
in new rails you can do something like that:
`params.require(:user).permit(:name, :email, :password)`

File: api/app/controllers/api/v1/auth_controller.rb [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+        user = User.find_by("LOWER(email) = ?", login_params[:email].to_s.downcase)
+
+        if user&.authenticate(login_params[:password])
+          tokens = JwtService.issue_tokens_for(user)
+          render json: {
+            access_token: tokens[:access_token],
+            refresh_token: tokens[:refresh_token]
+          }, status: :ok
+        else
+          render json: { error: "Invalid email or password" }, status: :unauthorized
```
Comment (le-murier) [inline api/app/controllers/api/v1/auth_controller.rb]: before, in some PRs, the team added error messages to the translations, you can do that too, just the English version. It’s not required and maybe won’t even be used in the app, as we most likely won’t need localization for those errors. But in general, people tend to avoid hard-coded string values that aren’t in .yml or some other files, since when the codebase grows, it becomes really painful to update or deal with just raw strings.

File: api/app/controllers/concerns/authenticatable.rb [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+  included do
+    rescue_from JwtService::ExpiredToken do |e| render json: { error: e.message }, status: :unauthorized end
+    rescue_from JwtService::InvalidToken do |e| render json: { error: e.message }, status: :unauthorized end
+    rescue_from ActiveRecord::RecordNotFound do
+      render json: { error: "User not found" }, status: :unauthorized
+    end
+  end
+
+  def current_user
```
Comment (le-murier) [inline api/app/controllers/concerns/authenticatable.rb]: I mean, it’s just a wrapper - it could be deleted, if you want it just for cleaner-looking code. Generally, in Rails, we tend to avoid simple wrappers around other variables/methods, as it's unnecessary code.

File: api/spec/models/user_spec.rb [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+    user = User.new
+
+    expect(user).to be_invalid
+    expect(user.errors[:name]).to be_present
+    expect(user.errors[:email]).to be_present
+    expect(user.errors[:password_digest]).to be_present
+  end
+
+  it "validates email format & uniqueness" do
+    User.create!(name: "A", email: "a@example.com", password: "secret123")
```
Comment (le-murier) [inline api/spec/models/user_spec.rb]: in 'it' block we generally have only expect(), this code you can add in before {}, or in let(:...)

File: api/spec/models/user_spec.rb [hunk 23]
```diff
@@ -0,0 +1,27 @@
+require "rails_helper"
+
+RSpec.describe User, type: :model do
+  it "validates presence and lengths" do
+    user = User.new
+
+    expect(user).to be_invalid
+    expect(user.errors[:name]).to be_present
+    expect(user.errors[:email]).to be_present
+    expect(user.errors[:password_digest]).to be_present
```
Comment (le-murier) [inline api/spec/models/user_spec.rb]: It is internal rails thing 'password_digest', it would be correct to test 'password'

File: api/spec/models/user_spec.rb [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+    expect(u2.errors[:email]).to include("has already been taken")
+  end
+
+  it "hashes password via has_secure_password" do
+    u = User.create!(name: "A", email: "x@y.com", password: "secret123")
+
+    expect(u.authenticate("secret123")).to eq(u)
+    expect(u.authenticate("nope")).to be_falsey
+  end
+end
```
Comment (le-murier) [inline api/spec/models/user_spec.rb:47]: Example +- how spec should look like (just example). Also some people make it like that: describe "#method_name" and then contexts of what you are testing:

```RUBY
require "rails_helper"

RSpec.describe User, type: :model do
  subject { described_class.new(attributes) }

  let(:attributes) { { name: "Alice", email: "alice@example.com", password: "secret123" } }

  describe "validations" do
    context "when fields are missing" do
      let(:attributes) { {} }

      it "is invalid and reports errors" do
        expect(subject).to be_invalid
        expect(subject.errors[:name]).to be_present
        expect(subject.errors[:email]).to be_present
        expect(subject.errors[:password]).to be_present
      end
    end

    context "when email is not unique" do
      before do
        User.create!(name: "Existing", email: "user@example.com", password: "secret123")
      end

      let(:attributes) { { name: "New", email: "USER@example.com", password: "secret123" } }

      it "is invalid due to case-insensitive email uniqueness" do
        expect(subject).to be_invalid
        expect(subject.errors[:email]).to include("has already been taken")
      end
    end
  end

  describe "#authenticate" do
    let!(:user) { User.create!(name: "Auth", email: "auth@example.com", password: "secret123") }

    it "returns user on correct password" do
      expect(user.authenticate("secret123")).to eq(user)
    end

    it "returns false on incorrect password" do
      expect(user.authenticate("wrongpass")).to be_falsey
    end
  end
end
```

File: api/spec/requests/api/v1/auth_spec.rb [hunk 25]
```diff
@@ -0,0 +1,80 @@
+require "rails_helper"
+
+RSpec.describe "Auth", type: :request do
+  before do
+    Role.find_or_create_by!(name: "USER")
+    Role.find_or_create_by!(name: "ADMIN")
+  end
+
+  describe "POST /api/v1/auth/register" do
```
Comment (le-murier) [inline api/spec/requests/api/v1/auth_spec.rb:8]: nice controller spec 🔥

File: api/spec/requests/api/v1/auth_spec.rb [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+    it "creates a user and returns tokens + user object" do
+      params = { name: "Jane", email: "jane@example.com", password: "password123" }
+      post "/api/v1/auth/register", params: params, as: :json
+
+      expect(response).to have_http_status(:created)
+      json = JSON.parse(response.body)
+      expect(json["access_token"]).to be_present
+      expect(json["refresh_token"]).to be_present
```
Comment (le-murier) [inline api/spec/requests/api/v1/auth_spec.rb]: can do something like one 'it' block verifying if status is fine, and second - all this fields to not over crumb single 'it' block

**Review by le-murier — COMMENTED** [review]
lgmt, minor comments

**Review by mehalyna — DISMISSED** [review]

File: api/.env.example (hunk 16 above)
Comment (PavloKilko) [reply api/.env.example:6]: Team doesn't know, but I think for now use of ENV-files will be much more simple for development and possible deployment.

File: api/app/controllers/api/v1/auth_controller.rb (hunk 17 above)
Comment (PavloKilko) [reply api/app/controllers/api/v1/auth_controller.rb]: Refactored:

+ Added PORO UserSerializer
+ Moved "business-logic" of the AuthController into separate services (/app/services/auth/login.rb & /app/**/register.rb)
+ Token generation is already exists in the JwtService as issue_tokens_for(user)

Comment (PavloKilko) [reply api/app/controllers/api/v1/auth_controller.rb]: Refactored: added into Role NAMES hashmap, so if some change will occur, there will an single source of change.

File: api/app/controllers/api/v1/auth_controller.rb (hunk 18 above)
Comment (PavloKilko) [reply api/app/controllers/api/v1/auth_controller.rb]: Fixed: added normalization on user model level, added constraint for user migration.

File: api/app/controllers/api/v1/auth_controller.rb (hunk 19 above)
Comment (PavloKilko) [reply api/app/controllers/api/v1/auth_controller.rb:25]: Based on the issue description, we're expecting json with plain params without grouping under the "block", just like this: `{"name": **, "email": **, "password": **}`

So, idk, didn't change it.

File: api/app/controllers/api/v1/auth_controller.rb (hunk 20 above)
Comment (PavloKilko) [reply api/app/controllers/api/v1/auth_controller.rb]: Fixed: reworked and added global API error handling with possible i18n. (Didn't create i18n for Ukrainian and I highly doubt that there exists importance to do this).

File: api/app/controllers/concerns/authenticatable.rb (hunk 21 above)
Comment (PavloKilko) [reply api/app/controllers/concerns/authenticatable.rb]: Fixed: changed on attr_reader.

File: api/spec/models/user_spec.rb (hunk 22 above)
Comment (PavloKilko) [reply api/spec/models/user_spec.rb]: Fixed: rewrited rspec for user model.

File: api/spec/models/user_spec.rb (hunk 23 above)
Comment (PavloKilko) [reply api/spec/models/user_spec.rb]: Fixed.

File: api/spec/models/user_spec.rb (hunk 24 above)
Comment (PavloKilko) [reply api/spec/models/user_spec.rb:47]: Fixed.

File: api/spec/requests/api/v1/auth_spec.rb (hunk 25 above)
Comment (PavloKilko) [reply api/spec/requests/api/v1/auth_spec.rb:8]: Thx, glad to hear it.

File: api/spec/requests/api/v1/auth_spec.rb (hunk 26 above)
Comment (PavloKilko) [reply api/spec/requests/api/v1/auth_spec.rb]: Fixed.

**Review by le-murier — DISMISSED** [review]

**Review by SomeTyraredGuy — DISMISSED** [review]

**Review by ruslanzhadko — APPROVED** [review]

**Review by SomeTyraredGuy — APPROVED** [review]

**Review by mehalyna — APPROVED** [review]
